LSP: semantic tokens support

jdrouhard opened the pull request:
This is a continuation of #15723. This will take into account all the feedback accumulated over the year(s) there and take advantage of newer core LSP functionality to hopefully provide a good baseline for semantic tokens support.

- [x] Determine if core should provide basic highlighting support
- [x] Fix document version synchronization
- [x] Fix documentation to match new module and API
- [x] Update tests for new functionality and API

gpanders commented on src/nvim/auevents.lua:
```diff
@@ -72,6 +72,7 @@ return {
     'InsertLeavePre',         -- just before leaving Insert mode
     'LspAttach',              -- after an LSP client attaches to a buffer
     'LspDetach',              -- after an LSP client detaches from a buffer
+    'LspSemanticTokens',      -- after an LSP semantic tokens request is processed
```
We are already using `LspRequest` as a User autocommand, maybe we should just promote that to a first class autocmd and make the autocommand pattern match on the request type.

So rather than a new `LspSemanticTokens` autocommand, one would use `autocmd LspRequest textDocument/semanticTokens/*`.

jdrouhard replied:
Can you direct a patterned autocmd to a specific buffer? Documentation (and reality) seems to imply you can’t mix them.

gpanders replied:
Correct, you cannot. The pattern needs to be a filename in order to do that.

However, we now have support for passing arbitrary data to Lua autocommand callbacks. So we could use that to pass the request to the autocommand callback. We already do this for `LspAttach` and `LspDetach`.

jdrouhard replied:
So you're suggesting adding a "pattern" to the arbitrary data table so consumers of the generic autocmd can distinguish one for semantic tokens? Something like:

```lua
api.nvim_exec_autocmds('LspResponse', {
  buffer = bufnr,
  data = {
    method = 'textDocument/semanticTokens', -- add /full? /full/*? Not sure what this should be exactly assuming range requests are added at some point 
    highlights = buf_state.highlights,
  }
})
```

`LspRequest` itself doesn't work because that's fired when requests are sent; not when responses are received. We could promote both for symmetry.

mfussenegger replied:
I don't think `LspRequest` is a good match here, it maps 1:1 to lsp requests, but the token event is synthesized.
If we're going to use autocmds I think we'll need a new dedicated event.

An alternative could be to move away from the extensiblity/semantic-token abstraction level, and instead go for a  high level LSP highlighting API/functionality. Without allowing access to tokens and also hiding the LSP semantic token semantics. E.g. something like `vim.lsp.hl_attach(client, buf)` that you can call within a `LspAttach` event and that keeps everything else implementation detail, without any other public API.

I'm not sure if that's better or worse, but it would focus the API more on the common case users are going to use it for, and plugins who really want to use the tokens could still use the lsp client's request functions directly to retrieve them.

> Determine if core should provide basic highlighting support

Yes I think it should.

jdrouhard replied:
Closing since this comment is no longer relevant to the design.

jdrouhard committed:
feat(lsp): initial support for semantic token highlighting

* credit to @smolck and @theHamsta for their contributions in laying the
  groundwork for this feature and for their work on some of the helper
  utility functions and tests

clason commented on runtime/lua/vim/lsp/semantic_tokens.lua:
```diff
@@ -0,0 +1,504 @@
+---
+---Finally, a redraw command is issued to force nvim to redraw the screen to
+---pick up changed highlight tokens.
+local function on_full_response(err, response, ctx, _)
```
mark as `---@private`

jdrouhard replied:
Yeah, was fixing gen_vimdoc.py as you were reviewing :)

clason commented on runtime/doc/lsp.txt:
```diff
@@ -1321,6 +1321,58 @@ save({lenses}, {bufnr}, {client_id})                 *vim.lsp.codelens.save()*
+==============================================================================
+Lua module: vim.lsp.semantic_tokens                      *lsp-semantic_tokens*
+
+                                           *vim.lsp.semantic_tokens.on_attach()*
```
These will be overwritten by the docstrings in your implementation.

Also, I don't think we should officially expose them. If you need to use them in other modules, that's ok, but we should still mark them as `---@private`. The idea is that nothing here is user-facing (although plugins may use the functions under a big, fat NO WARRANTY disclaimer).

clason reviewed (commented):
The doc changes

clason commented on runtime/lua/vim/lsp/semantic_tokens.lua:
```diff
@@ -0,0 +1,522 @@
+--- delete the semanticTokensProvider table from the {server_capabilities} of
+--- your client in your |LspAttach| callback or your configuration's
+--- `on_attach` callback.
+---
+--- <pre>
+---    client.server_capabilities.semanticTokensProvider = nil
```
```suggestion
--- <pre>
---  client.server_capabilities.semanticTokensProvider = nil
```

jdrouhard commented on runtime/doc/lsp.txt:
```diff
@@ -1321,6 +1321,51 @@ save({lenses}, {bufnr}, {client_id})                 *vim.lsp.codelens.save()*
+    Start the semantic token highlighting engine for the given buffer with the
+    given client.
+
+    NOTE: This is currently called automatically by |buf_attach_client|. To
+    opt-out of semantic highlighting with a server that supports it, you can
+    delete the semanticTokensProvider table from the {server_capabilities} of
+    your client in your |LspAttach| callback or your configuration's
+    `on_attach` callback.
```
This feature (semantic token highlighting) could very easily be opt-in instead of opt-out. API-wise, this now matches the way tree sitter works (start/stop methods) and seems like something that is safe to expose publicly from the beginning.

Making it opt in would simplify some of the changes in lsp.lua and wouldn't require a "hack" to disable it if someone doesn't want it. The explicit calls to start and stop in the lsp module could just disappear.

Opt-in is easy:

```lua
api.nvim_create_autocmd('LspAttach', {
    group = ...,
    callback = function(args) {
        vim.lsp.semantic_tokens.start(args.bufnr, args.data.client_id)
    }
})
```

When the client detaches from the buffer, the semantic token engine will automatically detach itself as well (the check in the `on_lines` callback), so this is all anyone would need in their config to enable it.

jdrouhard replied:
This is still an open question even though this comment is attached to an outdated diff.

jdrouhard commented on runtime/lua/vim/lsp/semantic_tokens.lua:
```diff
@@ -0,0 +1,573 @@
+  -- TODO: we want to keep as much of this API private as possible, but the main
+  -- lsp module needs access in its send_changes() function to be able to request
+  -- tokens on a buffer sync event. The alternative is to set up our own debounce
+  -- timer attached to on_lines, or wait for #15275.
```
This is the other part I'm not 100% sure on. Two things about this I'm not a huge fan of:

1. Requires a callback when the `textDocument/didChange` notification is sent. #15275 would "fix" this and no longer require a special call to the semantic token module inside the `send_changes()` call in the main lsp module. This also means we have to have an API function accessible outside this module that sends the semantic tokens request.
2. Sending a semantic tokens request for every didChange notification can be excessive.

I allude to it in the TODO here, but we can do this a couple different ways.

1. We can manage our own debounce timer that we can start in `on_lines`. I have an alternate branch that does this and it works well. We'd need a configuration option (per client? take it from cursor hold? not sure) for the actual debounce time. If we did this, I might also add a new autocmd for `InsertLeave` that cancels the timer and sends the request immediately.
2. We can add more autocmd hooks for things like `CursorHold`, `CursorHoldI`, etc and not worry about timers at all. This is a bit less flexible and could cause more delay than actually necessary in getting new tokens.

jdrouhard replied:
I opted for a configurable debounce timer that defaults to 200ms. It's configurable via an optional kwargs field in `start()`.  Feels cleaner to not need a specific hook in the didChange notification code.

Added `InsertLeave` to autocmds that send a new request so the debounce is cancelled and the request is sent immediately via n that event.

jdrouhard commented on runtime/lua/vim/lsp/semantic_tokens.lua:
```diff
@@ -0,0 +1,573 @@
+    callback = function()
+      request_all_tokens(bufnr, client_id)
+    end,
+  })
```
Instead of a `BufWinEnter` event, we could also just re-request tokens for all buffers for a client on a refresh request immediately. I don't really think one or the other is better. This current way is a little less chatter between nvim and the LSP server, but re-requesting all buffers immediately would be "snappier" for the user since switching back to a buffer would already have updated tokens and there'd be no delay waiting.

jdrouhard replied:
I think the `BufWinEnter` autocmd is the way to go, here. Spread the rpc/server load out if there's a refresh.

jdrouhard reviewed (commented):
@mfussenegger - thanks for taking the time to help with the API design and for looking things over.

@bfredl - I'd like to tag you for some input here as you have significant involvement with highlighting in general.

General question - should I be using `nvim_buf_add_highlight()`, or should I use the extmark API directly? The former doesn't let me specify a priority for these highlights, but I believe it's a slightly faster code path.

clason commented:
> General question - should I be using `nvim_buf_add_highlight()`, or should I use the extmark API directly? The former doesn't let me specify a priority for these highlights, but I believe it's a slightly faster code path.

`extmarks`, hands down. It should follow (and integrate with) treesitter highlighting as closely as possible. In particular, we need proper priority support.

theHamsta commented on runtime/lua/vim/lsp/semantic_tokens.lua:
```diff
@@ -0,0 +1,577 @@
+            state.namespace,
+            token.line,
+            token.start_col, {
+              hl_group = '@' .. token.type,
```
`@` were introduced for tree-sitter with some magic to support fallback with scopes. Would it make sense to add a different kind of magic highlight groups that is different from the tree-sitter ones and that might take into account token modifiers. Maybe similar to https://vscode-westeu.azurewebsites.net/api/language-extensions/semantic-highlight-guide#semantic-coloring-in-color-themes 

> `variable.readonly:java` is called a selector and has the form `(*|tokenType)(.tokenModifier)*(:tokenLanguage)?`

this could also be realized using the `@` fallbacks

clason replied:
No. We don't want to add yet a third type of highlight names. In particular, we _want_ to use the same highlight groups as for treesitter for tokens of the same name.

(And yes, one way of realizing modifiers similar to how VS Code does it is to support globs in highlight groups like `@foo.*.baz`, but I don't know if that's possible -- @bfredl ?)

jdrouhard replied:
We talked about this a bit offline. We do want to support modifiers (and potentially language specific groups), but we wanted to wait to add anything more extensible to a future PR after we get initial support in, which will simply link the LSP's token type to a same-named highlight group.

Modifiers might make more sense once we get the ability to do true highlight attribute combining, instead of trying to support an exponential explosion of highlight groups, or until we have something more akin to VSCode's "selector" that supports wildcards, like @clason was referring to.

theHamsta replied:
@clason VSCode does not support globs. The star just stands for having arbitrary modifiers in you mapping. E.g. `variable.readonly.global` matches a token of type `variable` with modifiers `readonly` and `global`. When the token is variable and has at least the two modifiers. VScode uses this notation just in configuration tables (or just consider token type for now). Neovim could do the same and then just use built-in highlight groups like `Type`. 

I was just confused why `@` was chosen. `@` from captures and names from LSP are two completely different sources and the same name might have different meanings. I would expect them to be mapped independently form each other. Color schemes would need to consider a union of tree-sitter captures and LSP token types. We could then at least provide a list of that union as the `@` groups that have a default mapping in built-in Nvim. So the planned the default experience with semantic token is that only the groups that have the same name as tree-sitter captures get a default mapping?

clason replied:
> @clason VSCode does not support globs. The star just stands for having arbitrary modifiers in you mapping.

I know; what I'm saying is that globs is how _we_ would probably realize that.

> I was just confused why @ was chosen. @ from captures and names from LSP are two completely different sources and the same name might have different meanings.

No. They refer to the same basic structures. More names is just more confusion.

> I would expect them to be mapped independently form each other. 

Why?

> Color schemes would need to consider a union of tree-sitter captures and LSP token types

Yes. Luckily the vast majority of tokens already exist as tree-sitter captures (with the same meaning).

And, again, we're not talking about modifiers yet in this PR.

jdrouhard replied:
To @theHamsta 's point, it does reduce flexibility by not having an LSP specific group in the mix.

The original way I wanted to do this was what he's talking about--create hlgroups for each semantic token type and default link them to tree sitter highlight groups where it makes sense. Generally the meanings are the same but forcing them together without any possible way to configure them differently seems unnecessarily prohibitive.

Doing this would give the same out of the box experience but gives more flexibility to colorscheme authors to tweak tree sitter and LSP toke s independently. If a colorscheme author doesn't care, the default links would still give semantic tokens meaningful highlights.

However, this is something we can explore in a future PR, as this one doesn't actually introduce any _new_ highlight groups that we may want to break later.

clason replied:
We discussed this; it increases maintenance (and documentation) surface for no apparent gain. Let's wait until someone comes up with a compelling(!) actual use case for distinguishing treesitter `@function` and LSP `function`.

jdrouhard replied:
The most compelling reason is simply that we'll need some links anyway, because not all semantic token type names directly map to a tree sitter capture name. If we need some, then it's actually more maintenance burden (IMO) to determine which token types need mappings because there isn't a tree sitter capture with the same name, and which do. Just create special LSP ones for all of them and map them all to something or nothing.

Again, future PR. Let's leave this one as is for now.

clason replied:
> The most compelling reason is simply that we'll need some links anyway, because not all semantic token type names directly map to a tree sitter capture name. 

No, we already have links: `@function` links to `Function`. We'd just add the missing ones to that (even if nvim-treesitter doesn't use them -- yet). The whole point is that we do _not_ want to add another layer of indirection, where LSP highlights map to treesitter highlights map to default (Vim) highlights.

jdrouhard replied:
I still disagree with this, but we can talk about it more on a future PR.

theHamsta replied:
> If you feel very strongly about this, use @function.lsp (which falls back to @function). But again, let's cross that bridge when we get to it.

This was the idea that I had first.  I had the feeling that it was less maintenance to not set nothing at all (or just have a simplistic mapping that can be overwritten). But probably just caring in the default experience for `@XXX` where `xxx` is already handled by tree-sitter mappings is not to bad. It doesn't have to be stable and people see something until someone convinces with something better.

rebelot commented:
Hi, I just wanted to share my implementation around handling of highlights.

The general idea is to:
1) resolve the highlight name based on server response to something like `"@keyword.modifier1.modifier2"`
2) call a utility that will create highlight default links recursively, i.e. `@keyword.modifier1.modifier2` links to `@keyword.modifier1`, which in turn links to `@keyword`. The utility can have a cache so that only the first occurrence of a particular token/modifier combination pays the price.
3) pass the hl name as is to extmarks

this allows to specify highlights "lazily", without the need of accounting for all combinations, but only the relevant ones.

### resolve highlight name and create default links
https://github.com/rebelot/dotfiles/blob/c3122f7067bbeb94e14329922f20dd9a46da2622/nvim/lua/lsp/semantic_tokens.lua#L205-L251

### parse highlight table map
https://github.com/rebelot/dotfiles/blob/c3122f7067bbeb94e14329922f20dd9a46da2622/nvim/lua/lsp/semantic_tokens.lua#L283-L294

### example highlight table map
https://github.com/rebelot/dotfiles/blob/c3122f7067bbeb94e14329922f20dd9a46da2622/nvim/lua/lsp/semantic_tokens.lua#L319-L360

Idk if this can be useful to any of you, just thought I would share this. Thanks for the great work :)

clason commented:
> @keyword.modifier1.modifier2 links to @keyword.modifier1, which in turn links to @keyword.

That's exactly how these highlight groups already behave in Neovim.

rebelot commented:
> > @keyword.modifier1.modifier2 links to @keyword.modifier1, which in turn links to @Keyword.
> 
> That's exactly how these highlight groups already behave in Neovim.

Indeed, but doesn't one need to explicitly create those links? Skimming through this PR I couldn't see any handling of token modifiers. The idea I wanted to share was just to create only the needed links at "runtime" instead of pre building all combinations beforehand. But again, I might have totally blundered what this is doing already.

clason commented:
> Indeed, but doesn't one need to explicitly create those links? Skimming through this PR I couldn't see any handling of token modifiers. The idea I wanted to share was just to create only the needed links at "runtime" instead of pre building all combinations beforehand. But again, I might have totally blundered what this is doing already.

For the semantic **tokens** (which this PR is about), we can and will just add default links for all groups (which do not already exist by virtue of them being used for treesitter).

For **modifiers** (which this PR is _not_ about), we haven't decided yet how we want to handle them. If there's any chance of doing it without having to define all possible combinations as separate highlight groups, that would be much preferable.

clason commented:
I _really_ want to keep modifiers out of this PR, as this feature has been derailed often enough.

When this is merged, we'll open a dedicated issue to discuss how these should behave based on 1. how modifiers are actually used in practice (e.g., what are the "relevant" combinations?) and 2. how this fits into the basic highlighting design in Neovim. I cannot stress enough that we do not want to add a special-case code path for these few highlights if it's at all possible to fit this into a general (improved) design.

mfussenegger commented on runtime/lua/vim/lsp/semantic_tokens.lua:
```diff
@@ -0,0 +1,621 @@
+  self:reset_timer()
+
+  for client_id, state in pairs(self.client_state) do
+    local client = vim.lsp.get_client_by_id(client_id)
```
I think there could be a race here where the `client` here becomes `nil`. For example when restarting the server, or if it crashes. 
Could probably skip the client in that case. The detach should run eventually and cleanup the state.

mfussenegger commented on runtime/lua/vim/lsp/semantic_tokens.lua:
```diff
@@ -0,0 +1,621 @@
+  end
+end
+
+--- |lsp-handler| for the method `textDocument/semanticTokens/full[/delta]`
+---
```
```suggestion
```

Comment is a bit misleading, the signature doesn't match that of `lsp-handler`. 
I'd remove the sentence. The description below explains what it does anyways.

jdrouhard replied:
It _started_ as the direct handler but I changed it so I could capture the version and pass it along to a member function. :-)

mfussenegger reviewed (commented):
Left a couple of comments. Looks pretty good to me so far.

jdrouhard commented on runtime/lua/vim/lsp/semantic_tokens.lua:
```diff
@@ -0,0 +1,621 @@
+
+---@private
+local function get_bit(n, k)
+  return _G.bit.band(_G.bit.rshift(n, k), 1)
```
Ah, the original PR actually had the Lua 5.1 version of this in here as well. I can add it back which should make this work for both.

mfussenegger replied:
Bundling is also already in progress: https://github.com/neovim/neovim/pull/21222

thetomato2 commented:
On windows in CLRF/DOS files I get this error:

![nvim_semantic_tokens_clrf_error](https://user-images.githubusercontent.com/56491112/204683434-66584205-4b95-4069-9867-ca542dc87d27.png)
I did know how to get to wherever those errors are logged lol. 
There are no issues in LF/UNIX files. 
Perusing the code I think it's [this line](https://github.com/neovim/neovim/blob/582cb3060dca984c94229ba8c8806f15da07607c/runtime/lua/vim/lsp/semantic_tokens.lua#L108) but idk, I don't have time to dive into this right now. I just assume no-one here is using windows and it's probably a trivial fix but if you need more info or someone to test just let  me know. And many thanks getting this feature along, its much needed when using C++.

theHamsta commented:
Could you add me and @smolck as co-authors for the first commit?

jdrouhard commented:
> Could you add me and @smolck as co-authors for the first commit?

I ended up rewriting this essentially from scratch but I can give you guys a credit for all the hard work you did laying the foundation in previous PRs! I'll probably reuse some of the test code and I used a helper function or two from the last PR.

Thanks for working on this feature and trying to push it through, I definitely used your branch for a solid year 😄

leaxoy commented:
Is this ready to merge, I can't wait to try this feature.

clason commented on runtime/lua/vim/lsp/semantic_tokens.lua:
```diff
@@ -0,0 +1,660 @@
+--- your client in your |LspAttach| callback or your configuration's
+--- `on_attach` callback.
+---
+--- <pre>
+--- client.server_capabilities.semanticTokensProvider = nil
+--- </pre>
```
```suggestion
--- <pre>lua
--- client.server_capabilities.semanticTokensProvider = nil
--- </pre>
```
(gives Lua syntax highlighting in generated docs)

jdrouhard commented:
@mfussenegger  @clason 

Sorry it took so long to get this across the finish line.

This PR still has the feature as opt-out if someone doesn't want it, instead of opt-in. I still think I slightly prefer opt-in instead. Currently, you opt out by nil'ing the `semanticTokensProvider` on server_capabilities in the client `on_attach` callback or `LspAttach` function. Feels a bit clunky.

OTOH, to opt-in, it's simply calling `vim.lsp.semantic_tokens.start(bufnr, client_id)` from the same `on_attach` or `LspAttach` function. Very easy to do and makes it easier to specify options (should we ever add more, like a client's highlighting priority override or something).

jdrouhard commented on runtime/lua/vim/lsp/semantic_tokens.lua:
```diff
@@ -0,0 +1,644 @@
+    highlighter = STHighlighter.new(bufnr)
+    highlighter.debounce = opts.debounce or 200
+  else
+    highlighter.debounce = math.max(highlighter.debounce, opts.debounce or 200)
```
Instead of taking the highest debounce of all attached clients in a buffer, we could also just use separate timers per client/per buffer (and that way they could be individually timed/configured). It would complicate things slightly since currently `send_request()` aggregates the clients and sends the request to all of them simultaneously.

In practice, this will almost never matter. I am not aware of any actual use-case for having multiple LSP servers provide semantic tokens at the same time.

clason replied:
>  I am not aware of any actual use-case for having multiple LSP servers provide semantic tokens at the same time.

When has that ever stopped anyone?

But you're right, having multiple language servers and possibly wanting more control over which one gets to paint the buffer is a convincing argument for making semantic highlighting opt-in. Let's go with your suggestion then (and add how to do this to the `news.txt` entry).

mfussenegger replied:
I'd still keep it opt-out instead of opt-in because it's non-intrusive and I think we should rather target the common use case instead of the niche use case. It's not ideal if we have to tell everyone to add a line to their LspAttach handler.

clason replied:
I don't have strong opinions either way; it's just slightly easier to add the `start()` mapping to `on_attach` or `LspAttached` than to edit the client capabilities for individual servers.

But if we keep it opt-out, do we want to expose (documented) the `start()` function at all? It's not really needed then, and keeping it private makes it easier to change things around later on (e.g., if that becomes necessary when adding modifiers into the mix).

Just some thoughts; to be clear, @mfussenegger has the last word here.

jdrouhard replied:
I'll leave everything as-is, for now, then, though this comment specifically was about the debounce timer being aggregated for all attached clients, instead of separated out :)

I'd like to leave start()/stop(). I think it's invaluable to be able to disable it (even if it's enabled for you automatically) for debugging purposes. Creating a simple toggler in my config was invaluable to be able to see the differences the semantic token highlighting was making, and I think there are probably other use cases people could come up with for being able to enable/disable the highlighting in a buffer without wholesale disabling the feature entirely.

Icelk commented:
I've been using this fork for some time now. When I enable spelling, documentation (only tested in Rust code) gets odd colouring. This also happens with COC: https://github.com/fannheyward/coc-rust-analyzer/issues/1083
Do you have any idea of the cause, and if this will be fixed?
Thanks!

jdrouhard committed:
refactor(test): create an lsp-specific helpers.lua file

* move the code for creating fake rpc servers there

mfussenegger commented on src/nvim/highlight_group.c:
```diff
@@ -264,6 +264,18 @@ static const char *highlight_init_both[] = {
   "default link @tag Tag",
+
+  // LSP semantic tokens
+  "default link @class Type",
+  "default link @struct Type",
+  "default link @enum Type",
+  "default link @enumMember Constant",
+  "default link @event Identifier",
+  "default link @interface Identifier",
+  "default link @modifier Identifier",
+  "default link @regexp SpecialChar",
+  "default link @typeParameter Type",
+  "default link @decorator Identifier",
```
@clason are these default links okay?

clason replied:
Looks reasonable enough to me, but I personally don't rely on the default links, so no strong feelings either way. @gpanders?

(Before 0.9, we probably want to take a good hard look at the default links (or colors) for the new treesitter and lsp groups anyway.)

clason replied:
Meaning, it shouldn't block this. As long as the links are there, we can always tweak them later (and probably will have to when modifiers enter the mix).

Oh, but:

> "default link @struct Type",

I'd prefer to rename the already existing `@structure` group instead -- having both is needlessly confusing.

clason commented on test/functional/plugin/lsp/helpers.lua:
```diff
@@ -0,0 +1,178 @@
+
+function M.clear_notrace()
+  -- problem: here be dragons
+  -- solution: don't look for dragons to closely
```
```suggestion
  -- solution: don't look too closely for dragons
```

clason commented on src/nvim/highlight_group.c:
```diff
@@ -258,12 +258,24 @@ static const char *highlight_init_both[] = {
   "default link @type.definition Typedef",
   "default link @storageclass StorageClass",
-  "default link @structure Structure",
+  "default link @struct Structure",
```
```suggestion
```

theHamsta commented:
Is it just me or does the new semantic highlighting significantly slow down the editor on large files (e.g. https://github.com/theHamsta/VideoProcessingFramework/blob/2cb837982155328bb8ac9f8b6355e4e20eb74644/src/TC/third_party/nvEncodeAPI.h#L3372)? I didn't have this experience with the previous versions of semantic highlighting.

jdrouhard commented:
> Is it just me or does the new semantic highlighting significantly slow down the editor on large files (e.g. https://github.com/theHamsta/VideoProcessingFramework/blob/2cb837982155328bb8ac9f8b6355e4e20eb74644/src/TC/third_party/nvEncodeAPI.h#L3372)? I didn't have this experience with the previous versions of semantic highlighting. Might be a config issue on my side or a subjective feeling. I tried to switch a few times back and before that commit.

When scrolling, or when inserting new text, or...? Theoretically this PR should be snappier since it only adds extmarks for visible regions as you move around, but if the list of tokens is large enough, the binary search it performs to find the "start" highlight on every redraw might be slower than just adding all the extmarks at once.  We could experiment with not using a decorator or storing the list of highlight tokens and simply adding every extmark as the response is processed--this would simplify part of the implementation as well.

theHamsta commented:
@jdrouhard I might have forgotten to run `:PackerCompile`, so I had the old and the new triggers for semantic highlighting running (even though I removed them from my config). It was when I was editing. It seems to be better now after a `:PackerCompile`.

nfrid commented:
for the modifiers I may propose a very dumb, but imo fairly handy approach: we can just add two more priorities like `semantic_tokens_modifiers = 126` & `semantic_tokens_own_modifiers = 127` and create extmarks `@modifier` & `@token.modifier` accordingly, so we can decide on how the symbol should be highlighted. It would provide us a simple way to distinguish, for example, variable, readonly **and** readonly variable as well w/out any weird APIs to handle it..

clason commented:
That won't work, because extmarks can't be combined (as was discussed in previous comments). You would need to create all possible combinations of tokens and modifiers, which is prohibitive.

Again: modifiers are off-topic for this PR. We'll handle it with a dedicated issue/PR, where we welcome input.
